### PR TITLE
[NwnPuJ91] Rewrite tests for failures in a safer way

### PR DESCRIPTION
**Pull request by Lojjs**
Tests with '(expected = QueryExecutionException.class)' or similar can
fail with the correct exception for unrelated reasons before coming to the assertion, making the test green even if it is broken. Also ignoreExceptions() feels dangerous to have in test setups 

Builds on top of https://github.com/neo4j/apoc/pull/240

### CODE REVIEW

File: core/src/test/java/apoc/atomic/AtomicTest.java [hunk 1]
```diff
@@ -161,19 +165,31 @@ public void testRemoveLastItemArray(){
...
-		assertThat(ArrayUtils.toObject(ages), Matchers.arrayContaining(40L, 50L, 60L));
+
+		QueryExecutionException e = assertThrows(QueryExecutionException.class,
+				() ->  testCall(db, "MATCH (n:Person {name:'Tom'}) CALL apoc.atomic.remove(n,$property,$position,$times) YIELD container RETURN count(*)",
+						map("property","age","position",5,"times",5), (r) -> {})
+
+		);
+		Throwable except = ExceptionUtils.getRootCause(e);
+        assertTrue(except instanceof RuntimeException);
+		assertEquals("Attention your position out of range or higher than array length, that is 3", except.getMessage());
```
Comment (gem-neo4j) [inline core/src/test/java/apoc/atomic/AtomicTest.java:179]: Nothing to do with this PR, but maybe we need to fix the English of our APOC Error messages sometime 🙈

File: core/src/test/java/apoc/load/LoadJsonTest.java [hunk 2]
```diff
@@ -321,19 +322,16 @@ public void testLoadJsonWithPathOptions() throws Exception {
...
-        }
+    @Test
+    public void testLoadJsonByUrlInConfigFileWrongKey() {
+        QueryExecutionException e = assertThrows(QueryExecutionException.class,
+                () -> testResult(db, "CALL apoc.load.json($key)",map("key","foo"), (r) -> r.hasNext())
+        );
+        Throwable except = ExceptionUtils.getRootCause(e);
+        assertTrue(except instanceof IOException);
+        final String message = except.getMessage();
+        assertTrue(message.startsWith("Cannot open file "));
```
Comment (gem-neo4j) [inline core/src/test/java/apoc/load/LoadJsonTest.java:333]: Do we know what is in the middle of this string?

File: core/src/test/java/apoc/path/PathsTest.java [hunk 3]
```diff
@@ -51,9 +51,14 @@ public void slice() throws Exception {
...
     }
 
-    @Test(expected = QueryExecutionException.class)
-    public void combineFail() throws Exception {
-        TestUtil.testCall(db, "MATCH p1 = (a:A)-[:NEXT]->(b:B), p2 = (c:C)-[:NEXT]->() RETURN apoc.path.combine(p1,p2) AS p", (row) -> fail("Should have failed"));
+    @Test
+    public void combineFail() {
+        QueryExecutionException e = assertThrows(QueryExecutionException.class,
+                () -> TestUtil.testCall(db, "MATCH p1 = (a:A)-[:NEXT]->(b:B), p2 = (c:C)-[:NEXT]->() RETURN apoc.path.combine(p1,p2) AS p",
+                        (row) -> fail("Should have failed"))
```
Comment (gem-neo4j) [inline core/src/test/java/apoc/path/PathsTest.java:58]: Do we need this row check?

File: core/src/test/java/apoc/periodic/PeriodicTest.java [hunk 4]
```diff
@@ -542,53 +547,62 @@ private long tryReadCount(int maxAttempts, String statement, long expected) thro
...
-    @Test(expected = QueryExecutionException.class, timeout = 1000)
+    @Test
     public void testIterateQueryFailInvalidConcurrency() {
         final String query = "CALL apoc.periodic.iterate('UNWIND range(0, 10) AS x RETURN x', " +
                 "'RETURN x', " +
                 "{concurrency:0 ,parallel:true})";
-        testFail(query);
+
+        QueryExecutionException e = assertThrows(QueryExecutionException.class,
+                () -> testCall(db, query, row -> fail("The test should fail but it didn't"))
```
Comment (gem-neo4j) [inline core/src/test/java/apoc/periodic/PeriodicTest.java:590]: Do we need this comment on the row?

File: core/src/test/java/apoc/util/UtilTest.java [hunk 5]
```diff
@@ -79,14 +80,12 @@ public void mergeNullMaps() {
...
+    @Test
     public void testValidateQuery() {
-        try {
-            Util.validateQuery(db, "Match (n) return m");
-        } catch (QueryExecutionException e) {
-            assertTrue(e.getMessage().contains("Variable `m` not defined"));
-            throw e;
-        }
+        QueryExecutionException e = assertThrows(QueryExecutionException.class,
+                () -> Util.validateQuery(db, "Match (n) return m")
```
Comment (gem-neo4j) [inline core/src/test/java/apoc/util/UtilTest.java:86]: Not really a part of the PR, but it kinda bothers me our Cypher is written this way, would be nice to have it match the style guide, but super low prio I guess 🤷‍♀️

**Review by gem-neo4j — APPROVED** [review]

File: core/src/test/java/apoc/atomic/AtomicTest.java (hunk 1 above)
Comment (Lojjs) [reply core/src/test/java/apoc/atomic/AtomicTest.java:179]: I fixed this one as it was indeed very bad English

File: core/src/test/java/apoc/load/LoadJsonTest.java (hunk 2 above)
Comment (Lojjs) [reply core/src/test/java/apoc/load/LoadJsonTest.java:333]: It contains the full path which is user/CI dependent (in my case is  /Users/louise/Documents/neo4j-apoc-core/core/) so there is a good reason it doesn't check for the whole message

Comment (gem-neo4j) [reply core/src/test/java/apoc/load/LoadJsonTest.java:333]: Ah that makes sense :) Thanks

File: core/src/test/java/apoc/util/UtilTest.java (hunk 5 above)
Comment (Lojjs) [reply core/src/test/java/apoc/util/UtilTest.java]: Fixed this one, if you want to do it for more places you can create a DayOFCleanup card

**Pull request merged**